### PR TITLE
Add `tctl create cap` for dynamically configuring cluster auth preference

### PR DESCRIPTION
**Pull request by andrejtokarcik**
In accordance with [RFD 16: Dynamic Configuration](https://github.com/gravitational/teleport/pull/5147).

### CODE REVIEW

**Comment by andrejtokarcik** [comment]
@Joerger Could you have a look at my additions to `api/types/constants.go`? I'm not clear about the intended difference between `api/types/constants.go` and `api/constants/constants.go`.

File: lib/auth/init_test.go [hunk 1]
```diff
@@ -158,85 +158,124 @@ func TestBadIdentity(t *testing.T) {
...
+	defer authServer.Close()
 
-	bk, err := lite.New(context.TODO(), backend.Params{"path": tempDir})
+	// Verify the stored auth preference is labelled as originating from config
+	// file.
+	storedAuthPref, err := authServer.GetAuthPreference()
 	require.NoError(t, err)
+	require.True(t, storedAuthPref.IsFromConfigFile())
+	defaultSecondFactor := services.DefaultAuthPreference().GetSecondFactor()
+	require.NotEqual(t, defaultSecondFactor, storedAuthPref.GetSecondFactor())
```
Comment (awly) [inline lib/auth/init_test.go:207]: nit: do `cmp.Diff` on the whole `AuthPreference`, here and below

File: lib/auth/init_test.go [hunk 2]
```diff
@@ -158,85 +158,124 @@ func TestBadIdentity(t *testing.T) {
...
+
+// TestAuthPreferenceResetDynamicToConfigFile ensures that a dynamically
+// configured AuthPreference can be reset back to a config-file preference.
+func TestAuthPreferenceResetDynamicToConfigFile(t *testing.T) {
+	// Simulate a server with auth preference set dynamically.
+	var err error
+	conf := setupConfig(t)
+	conf.AuthPreference, err = services.NewAuthPreference(services.AuthPreferenceSpecV2{
+		SecondFactor: constants.SecondFactorOff,
 	})
```
Comment (awly) [inline lib/auth/init_test.go:231]: isn't this setting up `AuthPreference` from config file?

File: lib/auth/init_test.go [hunk 3]
```diff
@@ -158,85 +158,124 @@ func TestBadIdentity(t *testing.T) {
...
-func TestClusterID(t *testing.T) {
-	tempDir := t.TempDir()
+// TestAuthPreferenceResetConfigFileToDefaults ensures that a config-file
+// AuthPreference can be reset back to defaults.
+func TestAuthPreferenceResetConfigFileToDefaults(t *testing.T) {
+	// Simulate a server with auth preference from config file.
+	var err error
+	conf := setupConfig(t)
+	conf.AuthPreference, err = services.NewAuthPreferenceFromConfigFile(services.AuthPreferenceSpecV2{
+		SecondFactor: constants.SecondFactorOff,
```
Comment (awly) [inline lib/auth/init_test.go:194]: set this to some other value, distinct from the one in default preference

File: lib/auth/init_test.go [hunk 4]
```diff
@@ -158,85 +158,124 @@ func TestBadIdentity(t *testing.T) {
...
+}
+
+// TestAuthPreferenceResetDynamicToConfigFile ensures that a dynamically
+// configured AuthPreference can be reset back to a config-file preference.
+func TestAuthPreferenceResetDynamicToConfigFile(t *testing.T) {
+	// Simulate a server with auth preference set dynamically.
+	var err error
+	conf := setupConfig(t)
+	conf.AuthPreference, err = services.NewAuthPreference(services.AuthPreferenceSpecV2{
+		SecondFactor: constants.SecondFactorOff,
```
Comment (awly) [inline lib/auth/init_test.go:230]: set this to some other value, distinct from the one in default preference

File: lib/auth/init_test.go [hunk 5]
```diff
@@ -158,85 +158,124 @@ func TestBadIdentity(t *testing.T) {
...
-		DataDir:        t.TempDir(),
-		HostUUID:       "00000000-0000-0000-0000-000000000000",
-		NodeName:       "foo",
-		Backend:        bk,
-		Authority:      testauthority.New(),
-		ClusterConfig:  services.DefaultClusterConfig(),
-		ClusterName:    clusterName,
-		StaticTokens:   services.DefaultStaticTokens(),
-		AuthPreference: authPreference,
+	// Attempt to reset back to default should be unsuccessful.
```
Comment (awly) [inline lib/auth/init_test.go:245]:
```suggestion
	// Attempt to reset back to default should be a no-op.
```

File: lib/auth/init_test.go [hunk 6]
```diff
@@ -158,85 +158,124 @@ func TestBadIdentity(t *testing.T) {
...
+	// defaults.
+	storedAuthPref, err = authServer.GetAuthPreference()
+	require.NoError(t, err)
+	require.True(t, storedAuthPref.IsFromDefaults())
+	require.Equal(t, defaultSecondFactor, storedAuthPref.GetSecondFactor())
+}
+
+// TestAuthPreferenceResetDynamicToConfigFile ensures that a dynamically
+// configured AuthPreference can be reset back to a config-file preference.
+func TestAuthPreferenceResetDynamicToConfigFile(t *testing.T) {
```
Comment (awly) [inline lib/auth/init_test.go:225]: add a test for the opposite - attempting to set dynamic preference when a static one is set

**Comment by Joerger** [comment]
> @Joerger Could you have a look at my additions to `api/types/constants.go`? I'm not clear about the intended difference between `api/types/constants.go` and `api/constants/constants.go`.

Constants that are used by the `api` module need to be in `api/constants` to prevent `api` from importing the main `teleport` module along with its 100+ dependencies. That's the only distinction, so you used it correctly.

File: api/types/constants.go [hunk 7]
```diff
@@ -225,6 +225,20 @@ const (
...
+	// that the resource originates from.
+	OriginLabel = "origin"
+
+	// OriginConfigFile is an origin value indicating that the resource is
+	// derived from static configuration.
+	OriginConfigFile = "config-file"
+
+	// OriginConfigFile is an origin value indicating that the resource was
+	// constructed as a default value.
+	OriginDefaults = "defaults"
```
[the remark this answers is not in the document]
Comment (andrejtokarcik) [reply api/types/constants.go:239]: No, any other string (including empty) is considered to indicate dynamic configuration.

File: api/types/resource.go [hunk 8]
```diff
@@ -67,6 +67,16 @@ type ResourceWithSecrets interface {
...
 }
 
+// ResourceWithOrigin provides information on the origin of the resource
+// (config-file, defaults, ...).
+type ResourceWithOrigin interface {
+	Resource
+	// IsFromConfigFile returns true if the resource originates from config file.
+	IsFromConfigFile() bool
+	// IsFromConfigFile returns true if the resource originates from defaults.
+	IsFromDefaults() bool
```
[the remark this answers is not in the document]
Comment (andrejtokarcik) [reply api/types/resource.go:77]: It was my intention to *not* make the interface universal in this sense to make sure our code always handles only those origin values that have an explicitly defined meaning. Since `Orgin()` would return just a plain string it could happen to be compared to an unrelated string value, either in an ad hoc manner (which we should discourage) or by accident (which we should prevent as well).

File: lib/auth/init.go [hunk 9]
```diff
@@ -284,11 +284,19 @@ func Init(cfg InitConfig, opts ...ServerOption) (*Server, error) {
...
 		return nil, trace.Wrap(err)
 	}
-	log.Infof("Updating cluster configuration: %v.", cfg.AuthPreference)
+	if updateAuthPref {
+		err = asrv.SetAuthPreference(cfg.AuthPreference)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		log.Infof("Updating auth preference: %v.", cfg.AuthPreference)
+	}
```
[the remark this answers is not in the document]
Comment (andrejtokarcik) [reply lib/auth/init.go:299]: If you mean to create a function that would handle this logic generically for any `ResourceWithOrigin` then I recall I tried but all my attempts failed due to Go's type system limitations, in particular its [lack of covariant types](https://golang.org/doc/faq#covariant_types).

If you mean to put this snippet into an `AuthPreference`-specific helper function (that would still need to be replicated for other `ResourceWithOrigin` types) then sure I can do that.

File: api/types/resource.go [hunk 10]
```diff
@@ -174,6 +184,11 @@ func (m *Metadata) Expiry() time.Time {
 	return *m.Expires
 }
 
+// Origin returns the origin value of the resource.
+func (m *Metadata) Origin() string {
+	return m.Labels[OriginLabel]
```
[the remark this answers is not in the document]
Comment (andrejtokarcik) [reply api/types/resource.go]: The original code should be OK. A nil map should return nil value for any key. It' assignment that's panic inducing.

I'm not sure about adding checks where they're not genuinely needed so maybe I'll revert your suggestion.

File: lib/auth/init.go [hunk 11]
```diff
@@ -477,6 +485,24 @@ func Init(cfg InitConfig, opts ...ServerOption) (*Server, error) {
 	return asrv, nil
 }
 
+// shouldUpdateResource determines whether the candidate resource
+// should be used to replace the currently stored resource.
+//
+// For more details, see Section "Auth server initialization"
+// in rfd/0016-dynamic-configuration.md
+func shouldUpdateResource(candidate services.ResourceWithOrigin, stored services.ResourceWithOrigin, getErr error) (bool, error) {
```
[the remark this answers is not in the document]
Comment (andrejtokarcik) [reply lib/auth/init.go]: Yeah but then the `trace.IsNotFound` would have to be evaluated outside of this function, right? My thinking was to make this function as generic as possible so that other `ResourceWithOrigin` can do without as less code duplication as possible.

File: lib/auth/init_test.go (hunk 2 above)
Comment (andrejtokarcik) [reply lib/auth/init_test.go:231]: I don't think so. `NewAuthPreference` creates the resource with no labels.

File: tool/tctl/common/resource_command.go [hunk 12]
```diff
@@ -81,17 +82,19 @@ Same as above, but using JSON output:
...
+		services.KindGithubConnector:       rc.createGithubConnector,
+		services.KindCertAuthority:         rc.createCertAuthority,
+		services.KindClusterAuthPreference: rc.createAuthPreference,
 	}
 	rc.config = config
 
 	rc.createCmd = app.Command("create", "Create or update a Teleport resource from a YAML file")
 	rc.createCmd.Arg("filename", "resource definition file, empty for stdin").StringVar(&rc.filename)
 	rc.createCmd.Flag("force", "Overwrite the resource if already exists").Short('f').BoolVar(&rc.force)
+	rc.createCmd.Flag("confirm", "Confirm an unsafe or temporary resource update").Hidden().BoolVar(&rc.confirm)
```
[the remark this answers is not in the document]
Comment (andrejtokarcik) [reply tool/tctl/common/resource_command.go:97]: Users are meant to use it only when they are asked to do so as an additional confirmation of their intent. It's counter-productive to show it in the CLI help.

File: lib/auth/init_test.go (hunk 3 above)
Comment (andrejtokarcik) [reply lib/auth/init_test.go:194]: `SecondFactorOff` *is* distinct from the default one which is `constants.SecondFactorOTP`...?

File: lib/auth/init_test.go (hunk 1 above)
Comment (andrejtokarcik) [reply lib/auth/init_test.go:207]: That doesn't really work because of the metadata changes that are involved.

File: tool/tctl/common/resource_command.go [hunk 13]
```diff
@@ -404,6 +406,39 @@ func (rc *ResourceCommand) createUser(client auth.ClientI, raw services.UnknownR
 	return nil
 }
 
+// createAuthPreference implements `tctl create cap.yaml` command.
+func (rc *ResourceCommand) createAuthPreference(client auth.ClientI, raw services.UnknownResource) error {
```
[the remark this answers is not in the document]
Comment (andrejtokarcik) [reply tool/tctl/common/resource_command.go:410]: Actually, none of these checks are really meant to be guarantees. They are mere guidance which follows from the fact that the `origin` label can be manually assigned any value in the resource YAML (including `config-file` that'd override anything in place) so it doesn't make much sense to verify them on the auth service side, cf. [the RFD](https://github.com/gravitational/teleport/blob/master/rfd/0016-dynamic-configuration.md#implementation):
> Note that the `origin` label is not reserved or protected in any special way by the system and so it can be modified by the same means as other labels. User activity can disrupt the intended semantics and cause the label's value to cease to be representative of the actual origin or values of a resource.

Which is in turn based on [this discussion](https://github.com/gravitational/teleport/pull/5147#issuecomment-749715119):
> That's the trade-off I am OK to make. If user overwrites the labels set by the system, they may break the system. Kubernetes follows the similar approach by marking owner reference in the object controller creates. They set system labels and document their usage. If you update the system labels, the system may misbehave.

File: api/types/resource.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (awly) [reply api/types/resource.go]: How about returning a typed string?
`type Origin string` and make the constants in `api/types/constants.go` use that.
That way you'd need to explicitly convert an unrelated string for comparison.

File: lib/auth/init.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (awly) [reply lib/auth/init.go:299]: I meant an `AuthPreference`-specific function.
This `func Init` is already massive and hard to follow

File: lib/auth/init_test.go (hunk 6 above)
Comment (awly) [reply lib/auth/init_test.go:225]: Ping on this

File: tool/tctl/common/resource_command.go [hunk 14]
```diff
@@ -404,6 +406,39 @@ func (rc *ResourceCommand) createUser(client auth.ClientI, raw services.UnknownR
...
+		return trace.AlreadyExists("non-default cluster auth preference already exists")
+	}
+
+	managedByStatic := storedAuthPref.IsFromConfigFile()
+	if !rc.confirm && managedByStatic {
+		return trace.Errorf("This resource is managed by static configuration. " +
+			"We recommend removing configuration from teleport.yaml, " +
+			"restarting the servers and trying this command again.\n\n" +
+			"If you would still like to proceed, re-run the command with both --force " +
+			"and --confirm flags.")
```
[the remark this answers is not in the document]
Comment (awly) [reply tool/tctl/common/resource_command.go:432]: Ping on this

File: lib/auth/init_test.go (hunk 1 above)
Comment (awly) [reply lib/auth/init_test.go]: Are there many metadata changes?
You can use https://pkg.go.dev/github.com/google/go-cmp/cmp/cmpopts#IgnoreFields to skip stuff like timestamps

File: tool/tctl/common/resource_command.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (awly) [reply tool/tctl/common/resource_command.go:410]: I disagree with that choice, but let's go with the approved design.
Please leave a comment here with those links so that the next reader can understand the decision.

**Comment by andrejtokarcik** [comment]
Two new features: If a reserved origin (as of now `config-file` or `defaults`) is used in the user resource YAML, an error similar to the following is returned:
```
ERROR: origin "config-file" is reserved for system use
```

Secondly, I made the permissions required for dynamically configuring a resource more granular:
https://github.com/gravitational/teleport/blob/923ad00fce4cee2efe086649283cf390cda247df/lib/auth/auth_with_roles.go#L2656-L2667
* Both `tctl create` (i.e. when the stored resource being overwritten is the default) and `tctl create --force` (i.e. when what's stored has already been configured *dynamically*) require `VerbUpdate`.
* `tctl create --force --confirm` (i.e. when the stored resource is from a config file) additionally requires `VerbCreate` (e.g. our Cloud users most probably shouldn't be granted this privilege).
* The `VerbDelete` permission is mentioned in anticipation of the [proposed `tctl rm` semantics](https://github.com/gravitational/teleport/pull/5673).

@klizhentas Would this be acceptable? If so I'll update the RFD with these rules.

File: lib/auth/init.go [hunk 15]
```diff
@@ -483,6 +482,49 @@ func Init(cfg InitConfig, opts ...ServerOption) (*Server, error) {
 	return asrv, nil
 }
 
+func setAuthPreference(asrv *Server, newAuthPref services.AuthPreference) error {
```
Comment (awly) [inline lib/auth/init.go:485]: nit: `initSetAuthPreference` to prevent someone accidentally calling it from a non-init codepath

File: lib/auth/init.go [hunk 16]
```diff
@@ -483,6 +482,49 @@ func Init(cfg InitConfig, opts ...ServerOption) (*Server, error) {
...
+	}
+	if stored == nil || stored.Origin() == types.OriginConfigFile || stored.Origin() == types.OriginDefaults {
+		return true, nil
+	}
+	if candidate.Origin() == types.OriginConfigFile {
+		// Log a warning when about to overwrite a dynamically configured resource.
+		log.Warnf("Stored %v resource with origin %q is to be discarded in favor of resource with origin %q.",
+			stored.GetKind(), stored.Origin(), candidate.Origin())
+		return true, nil
+	}
```
Comment (awly) [inline lib/auth/init.go:524]: this should `return false, nil`
as a Cloud user, I'd want to call `tctl create mydynamicpreference.yaml` and not have it wiped out when the auth server restarts next time

unless I'm misunderstanding the design, this whole function should be `return candidate.Origin() == types.OriginConfigFile || candidate.Origin() == types.OriginDefaults`

File: tool/tctl/common/resource_command.go [hunk 17]
```diff
@@ -742,3 +774,7 @@ func UpsertVerb(exists bool, force bool) string {
 		return "unknown"
 	}
 }
+
+const managedByStaticMsg = `This resource is managed by static configuration. We recommend removing configuration from teleport.yaml, restarting the servers and trying this command again.
+
+If you would still like to proceed, re-run the command with both --force and --confirm flags.`
```
Comment (awly) [inline tool/tctl/common/resource_command.go:780]: nit: inline this error message above

File: lib/auth/auth_with_roles.go [hunk 18]
```diff
@@ -2648,3 +2652,16 @@ func NewAdminAuthServer(authServer *Server, sessions session.Service, alog event
...
+func verbsForResourceWithOrigin(stored, candidate types.ResourceWithOrigin) []string {
+	verbs := []string{types.VerbUpdate}
+	if stored.Origin() == types.OriginConfigFile {
+		verbs = append(verbs, types.VerbCreate)
+	}
+	if candidate.Origin() == types.OriginDefaults {
+		verbs = append(verbs, types.VerbDelete)
+	}
+	return verbs
+}
```
[the remark this answers is not in the document]
Comment (andrejtokarcik) [reply lib/auth/auth_with_roles.go:2667]: The `VerbDelete` condition is probably going away.

However we really need a distinguished privilege (whether `VerbCreate` or not) to prevent mere `VerbUpdate` users from using `tctl create --force --confirm` and thus overwriting (even if just temporarily) the config-file configuration. There should be a straightforward way to make the config file authoritative once it's specified.

I think it's okay for the verbs to have a slightly shifted meaning in the context of singleton resources (that don't really get created or deleted).

File: tool/tctl/common/resource_command.go (hunk 17 above)
Comment (andrejtokarcik) [reply tool/tctl/common/resource_command.go:780]: It should be okay to leave this as-is since there may be more dynamic config resources coming, all of which could then refer to this message?

File: lib/auth/init.go (hunk 16 above)
Comment (andrejtokarcik) [reply lib/auth/init.go:524]: Your suggestion would replace any stored dynamic configuration with the default one upon restart.

Comment (awly) [reply lib/auth/init.go:524]: NVM, I mis-undestood the RFD

File: lib/auth/auth_with_roles.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (awly) [reply lib/auth/auth_with_roles.go:2667]: modifying `AuthPreference` in any way seems like a privileged operation, whether it's replacing a static or dynamic value.

[the remark this answers is not in the document]
Comment (andrejtokarcik) [reply lib/auth/auth_with_roles.go:2667]: Yes, that's why `VerbUpdate` would be always needed.

However, replacing a piece of static configuration should require even more privileges: e.g. `VerbCreate` in addition to the "basic" `VerbUpdate` as proposed here.

BTW, the convention with our handlers is to require both `VerbCreate` and `VerbUpdate` for a set operation.

[the remark this answers is not in the document]
Comment (awly) [reply lib/auth/auth_with_roles.go:2667]: Alright, please remove `VerbDelete` though

**Review by awly — APPROVED** [review]

File: lib/auth/init.go [hunk 19]
```diff
@@ -484,6 +483,49 @@ func Init(cfg InitConfig, opts ...ServerOption) (*Server, error) {
 	return asrv, nil
 }
 
+func initSetAuthPreference(asrv *Server, newAuthPref services.AuthPreference) error {
+	storedAuthPref, err := asrv.GetAuthPreference()
+	if err != nil {
+		if !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+		storedAuthPref = nil
```
Comment (r0mant) [inline lib/auth/init.go:492]: Wouldn't `storedAuthPref` be `nil` if above method returned an error so this assignment is redundant?

File: lib/auth/apiserver.go [hunk 20]
```diff
@@ -2494,6 +2498,12 @@ func (s *APIServer) getServerID(r *http.Request) (string, error) {
 	return strings.TrimSuffix(role.Username, "."+clusterName), nil
 }
 
+// isReservedOrigin returns true if the given origin is reserved for system use
+// and must not be accepted within a user resource.
+func isReservedOrigin(origin types.OriginValue) bool {
```
Comment (r0mant) [inline lib/auth/apiserver.go:2503]: If we change the label to be namespaced as I suggested above, would it make sense to just prevent it to be set by users completely instead of just restricting allowed values?

File: lib/auth/auth_with_roles.go [hunk 21]
```diff
@@ -2740,3 +2744,13 @@ func NewAdminAuthServer(authServer *Server, sessions session.Service, alog event
 		sessions:   sessions,
 	}, nil
 }
+
+// verbsForResourceWithOrigin determines the verbs/actions required of a role
+// to replace the resource currently stored in the backend with the candidate.
+func verbsForResourceWithOrigin(stored, candidate types.ResourceWithOrigin) []string {
```
Comment (r0mant) [inline lib/auth/auth_with_roles.go:2750]: `candidate` isn't used by this function.

File: lib/auth/init.go [hunk 22]
```diff
@@ -484,6 +483,49 @@ func Init(cfg InitConfig, opts ...ServerOption) (*Server, error) {
...
+	return nil
+}
+
+// shouldInitReplaceResource determines whether the candidate resource should
+// be used to replace the stored resource during auth server initialization.
+//
+// For more details, see Section "Auth server initialization"
+// in rfd/0016-dynamic-configuration.md
+func shouldInitReplaceResource(stored, candidate types.ResourceWithOrigin) (bool, error) {
+	if candidate.Origin() != types.OriginConfigFile && candidate.Origin() != types.OriginDefaults {
```
Comment (r0mant) [inline lib/auth/init.go:514]: Nit: `Check()` method on the Origin type may be helpful.

File: lib/auth/init.go [hunk 23]
```diff
@@ -484,6 +483,49 @@ func Init(cfg InitConfig, opts ...ServerOption) (*Server, error) {
...
+		return false, trace.BadParameter("candidate origin must be either config-file or defaults (this is a bug)")
```
Comment (r0mant) [inline lib/auth/init.go:515]: Where does the `candidate` resource come from? If I'm reading correctly, it's coming from the configuration file? If so, users may see this error and it will be confusing to them.

File: lib/auth/init.go (hunk 22 above)
Comment (r0mant) [reply lib/auth/init.go:514]: Also, can `candidate` be nil here?

File: lib/auth/init.go [hunk 24]
```diff
@@ -484,6 +483,49 @@ func Init(cfg InitConfig, opts ...ServerOption) (*Server, error) {
...
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		log.Infof("Updating auth preference: %v.", newAuthPref)
+	}
+	return nil
+}
+
+// shouldInitReplaceResource determines whether the candidate resource should
+// be used to replace the stored resource during auth server initialization.
```
Comment (r0mant) [inline lib/auth/init.go:509]: I would explain when this is the case and why in this godoc. We can keep the link to RFD but having to read the design document to understand why the function does what it does isn't very convenient IMO.

File: lib/auth/init.go [hunk 25]
```diff
@@ -484,6 +483,49 @@ func Init(cfg InitConfig, opts ...ServerOption) (*Server, error) {
...
+//
+// For more details, see Section "Auth server initialization"
+// in rfd/0016-dynamic-configuration.md
+func shouldInitReplaceResource(stored, candidate types.ResourceWithOrigin) (bool, error) {
+	if candidate.Origin() != types.OriginConfigFile && candidate.Origin() != types.OriginDefaults {
+		return false, trace.BadParameter("candidate origin must be either config-file or defaults (this is a bug)")
+	}
+	if stored == nil || stored.Origin() == types.OriginConfigFile || stored.Origin() == types.OriginDefaults {
```
Comment (r0mant) [inline lib/auth/init.go:517]: Add comment explaining why the resource should be replaced in this case.

In fact, I would do if for each `if` statement here.

File: tool/tctl/common/resource_command.go [hunk 26]
```diff
@@ -406,6 +409,35 @@ func (rc *ResourceCommand) createUser(client auth.ClientI, raw services.UnknownR
...
+	}
+
+	exists := storedAuthPref.Origin() != types.OriginDefaults
+	if !rc.force && exists {
+		return trace.AlreadyExists("non-default cluster auth preference already exists")
+	}
+
+	managedByStaticConfig := storedAuthPref.Origin() == types.OriginConfigFile
+	if !rc.confirm && managedByStaticConfig {
+		return trace.Errorf(managedByStaticConfigMsg)
```
Comment (r0mant) [inline tool/tctl/common/resource_command.go:431]: trace.BadParameter? We don't usually use trace.Errorf directly anywhere.

File: api/types/constants.go [hunk 27]
```diff
@@ -228,8 +231,21 @@ const (
...
 )
 
-// WebSessionSubKinds lists subkinds of web session resources
-var WebSessionSubKinds = []string{KindAppSession, KindWebSession}
+type OriginValue string
+
+const (
+	// OriginLabel is a resource metadata label name used to identify a source
+	// that the resource originates from.
+	OriginLabel = "origin"
```
[the remark this answers is not in the document]
Comment (andrejtokarcik) [reply api/types/constants.go:239]: By "namespacing" you mean just adding a prefix in the label name? I'm not sure about that, this is how it's defined in the RFD. My impression was that @klizhentas wanted to keep it simple & allow the users to tinker with their own origins as well if they wanted to...?

File: lib/auth/apiserver.go (hunk 20 above)
Comment (andrejtokarcik) [reply lib/auth/apiserver.go:2503]: Has to be brought up with @klizhentas as it would be a modification of the RFD design.

File: lib/auth/init.go (hunk 22 above)
Comment (andrejtokarcik) [reply lib/auth/init.go:514]: The `nil` case should be excluded here although it could occur by a programmer mistake. `candidate` is what comes from the default configuration which is updated with what's specified in the config file. I'll add a `nil` check just to be sure.

File: lib/auth/init.go (hunk 23 above)
Comment (andrejtokarcik) [reply lib/auth/init.go:515]: See above. This is just a sanity check for the resource-aware resources to come with properly set origin values. The error should be triggered only during development.

File: api/types/constants.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (r0mant) [reply api/types/constants.go:239]: > By "namespacing" you mean just adding a prefix in the label name?

Yeah, like in my example above: `teleport.io/resource-origin`. It's a pretty common practice to separate reserved labels names from user labels. E.g. Kubernetes does that too with their `kubernetes.io/xxx` labels.

File: api/types/authentication.go [hunk 28]
```diff
@@ -170,6 +188,16 @@ func (c *AuthPreferenceV2) SetResourceID(id int64) {
 	c.Metadata.ID = id
 }
 
+// Origin returns the origin value of the resource.
+func (c *AuthPreferenceV2) Origin() string {
```
Comment (r0mant) [inline api/types/authentication.go:192]: Nit: `GetOrigin()`? :)

File: api/types/resource.go [hunk 29]
```diff
@@ -201,6 +227,11 @@ func (m *Metadata) CheckAndSetDefaults() error {
 		}
 	}
 
+	// Check the origin value.
+	if !utils.SliceContainsStr(append(OriginValues, ""), m.Origin()) {
+		return trace.BadParameter("invalid origin value: %q", m.Origin())
```
Comment (r0mant) [inline api/types/resource.go:232]: Nit: We usually try to give possible accepted values in the error message in such cases, if that'd make sense here.

**Review by r0mant — APPROVED** [review]

File: api/types/authentication.go (hunk 28 above)
Comment (andrejtokarcik) [reply api/types/authentication.go:192]: Aren't we following [official Go guidelines](https://golang.org/doc/effective_go#Getters) as much as possible/reasonable with new code?

File: api/types/resource.go (hunk 29 above)
Comment (andrejtokarcik) [reply api/types/resource.go]: Done, added a note to the message that this occurring amounts to a bug.

**Comment by highb** [comment]
@andrejtokarcik Is this config resource documented anywhere? I don't see it in [the resource guide](https://goteleport.com/docs/admin-guide/#resources) yet.

**Comment by andrejtokarcik** [comment]
@highb It isn't, as far as I can tell. I opened https://github.com/gravitational/teleport/pull/7548 extending the list of resources with all the dynamically configurable resources that should become available in Teleport 7.0.
